builtin/logical/ssh/path_role_create.go

```diff
@@ -0,0 +1,193 @@
+package ssh
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net"
+	"strings"
+
+	"github.com/hashicorp/vault/logical"
+	"github.com/hashicorp/vault/logical/framework"
+)
+
+func pathRoleCreate(b *backend) *framework.Path {
+	return &framework.Path{
+		Pattern: "creds/(?P<name>\\w+)",
+		Fields: map[string]*framework.FieldSchema{
+			"name": &framework.FieldSchema{
+				Type:        framework.TypeString,
+				Description: "name of the policy",
+			},
+			"username": &framework.FieldSchema{
+				Type:        framework.TypeString,
+				Description: "username in target",
+			},
+			"ip": &framework.FieldSchema{
+				Type:        framework.TypeString,
+				Description: "IP of the target machine",
+			},
+		},
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.WriteOperation: b.pathRoleCreateWrite,
+		},
+		HelpSynopsis:    pathRoleCreateHelpSyn,
+		HelpDescription: pathRoleCreateHelpDesc,
+	}
+}
+
+func (b *backend) pathRoleCreateWrite(
+	req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	roleName := d.Get("name").(string)
+	username := d.Get("username").(string)
+	ipRaw := d.Get("ip").(string)
+	if roleName == "" {
+		return logical.ErrorResponse("Invalid 'name'"), nil
+	}
+	if ipRaw == "" {
+		return logical.ErrorResponse("Invalid 'ip'"), nil
+	}
+
+	//find the role to be used for installing dynamic key
+	rolePath := "policy/" + roleName
+	roleEntry, err := req.Storage.Get(rolePath)
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving role: %s", err)
+	}
+	if roleEntry == nil {
+		return logical.ErrorResponse(fmt.Sprintf("Role '%s' not found", roleName)), nil
+	}
+	var role sshRole
+	if err := roleEntry.DecodeJSON(&role); err != nil {
+		return nil, err
+	}
+
+	if username == "" {
+		username = role.DefaultUser
+	}
+
+	//validate the IP address
+	ipAddr := net.ParseIP(ipRaw)
+	if ipAddr == nil {
+		return logical.ErrorResponse(fmt.Sprintf("Invalid IP '%s'", ipRaw)), nil
+	}
+	ip := ipAddr.String()
+
+	ipMatched := false
+	for _, item := range strings.Split(role.CIDR, ",") {
+		_, cidrIPNet, err := net.ParseCIDR(item)
+		if err != nil {
+			return logical.ErrorResponse(fmt.Sprintf("Invalid cidr entry '%s'", item)), nil
+		}
+		ipMatched = cidrIPNet.Contains(ipAddr)
+		if ipMatched {
+			break
+		}
+	}
+	if !ipMatched {
+		return logical.ErrorResponse(fmt.Sprintf("IP[%s] does not belong to role[%s]", ip, roleName)), nil
+	}
+
+	//fetch the host key to be used for installation
+	keyPath := "keys/" + role.KeyName
+	keyEntry, err := req.Storage.Get(keyPath)
+	if err != nil {
+		return nil, fmt.Errorf("Key '%s' not found error:%s", role.KeyName, err)
+	}
+	var hostKey sshHostKey
+	if err := keyEntry.DecodeJSON(&hostKey); err != nil {
+		return nil, fmt.Errorf("Error reading the host key: %s", err)
+	}
+
+	//store the host key to file. Use it as parameter for scp command
+	hostKeyFileName := "./vault_ssh_" + username + "_" + ip + "_shared.pem"
```

Comment:
I think these should use `fmt.Sprintf` for interpolation so they're easier to read / idiomatic. E.g.

    hostKeyFileName := fmt.Sprintf("./vault_ssh_%s_%s_shared.pem", username, ip)

I can do the interpolation in my head with that; a lot harder to keep track of all the `+` and `"` in my head. :(